### PR TITLE
fix: CalendarView selection near end of entire calendar

### LINKED ISSUE #17635: [CalendarView] Significant Gap Appears in Calendar When when reaching end of calendar
### Current behavior

After selecting a day at the end of the calendar and switching back to the year view, a large gap is present, disrupting the user experience and causing visual inconsistencies.

https://github.com/user-attachments/assets/250d6ff2-c146-4643-b912-27e8a7d80a7b

Screenshot 2024-07-22 at 11 39 45 AM

### Expected behavior

The calendar should smoothly transition between month and year views without any large gaps or layout issues.

### How to reproduce it (as minimally and precisely as possible)

Steps to Reproduce:

1. Open Gallery in any of the web browser
2. Navigate to Menu -> UI Components -> CalenderView
3. In the calendar reach to end of years
4. Select a day
5. Go back to years
6. Observe the huge blank gap.

### Workaround

_No response_

### Works on UWP/WinUI

None

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

WebAssembly

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by rajamatt**
GitHub Issue (If applicable): closes #17635 

## PR Type

What kind of change does this PR introduce?

- Bugfix

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarView_Partial.cs [hunk 1]
```diff
@@ -1645,17 +1645,45 @@ public void SetYearDecadeDisplayDimensions(int columns, int rows)
...
 			DateTime date)
 		{
-			int index = 0;
-
-			index = pHost.CalculateOffsetFromMinDate(date);
+			int index = pHost.CalculateOffsetFromMinDate(date);
 			global::System.Diagnostics.Debug.Assert(index >= 0);
 			global::System.Diagnostics.Debug.Assert(pHost.Panel is { });
+
+#if HAS_UNO_SKIA || __WASM__
```
Comment (rajamatt) [inline src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarView_Partial.cs:1652]: I can't find any usages of `HAS_UNO_SKIA`, what directive do we use for platform specific skia code?

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarView_Partial.cs:1652]: We'll need to add tests for this scenario.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarView_Partial.cs:1652]: > I can't find any usages of `HAS_UNO_SKIA`, what directive do we use for platform specific skia code?

`__SKIA__` should work. Though I think this is ported code, so unclear why it needs to change

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarView_Partial.cs:1652]: Also note that the current offset looks correct, but the issue is in the "blank" area. It should contain months from the previous year.

The repro steps in the issue showed that it rendered correctly one time, then it no longer renders correctly.


**Correct:**

image

**Wrong:**

image

Comment (rajamatt) [reply src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarView_Partial.cs]: After digging into this further, I'm not sure what fix/workaround would be best. 

This is the behavior for the UWP CalendarView:

WindowsCalendarView

If we compare to the current version @Youssef1313 you are right, the indices are correct it's just the space above that needs to be filled in on other platforms.

Here's where the bounds of the scope of items in the current calendar view are set. If we could decrease the `FirstVisibleIndexBase`, more items would be visible and would fill the gap of this issue.

https://github.com/unoplatform/uno/blob/4ad6ef3bbb9064af827bc2751281cb2e47e31812/src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarView_Partial.cs#L1840-L1841

These indices `FirstVisibleIndexBase` and `LastVisibleIndexBase` are set via code that bubbles up from here:

https://github.com/unoplatform/uno/blob/4ad6ef3bbb9064af827bc2751281cb2e47e31812/src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarLayoutStrategyImpl.cs#L183-L186

I thought I would be able to replicate the changes in this PR in the piece of code above but I've had no luck so far. Everytime we switch from the decade view to the year view the `FirstVisibleIndexBase` increases for some reason I haven't been able to find, leaving even less items visible.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarView_Partial.cs]: @dr1rrb Does that sound familiar?

Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarView_Partial.cs]: Not really ... I remember that we already did fixes for the month / year / decade switching, but I'm unable to remember what is was, we need to look at git history.

**Pull request closed without merging**